larreco/QuadVtx/QuadVtx_module.cc

```diff
@@ -0,0 +1,522 @@
+      for(const Pt2D& p: pts[view]) gpts->SetPoint(gpts->GetN(), p.z, p.x);
+    }
+  }
+
+  float minx = +1e9;
+  float maxx = -1e9;
+  float minz[3] = {+1e9, +1e9, +1e9};
+  float maxz[3] = {-1e9, -1e9, -1e9};
+  for(int view = 0; view < 3; ++view){
+    for(const Pt2D& p: pts[view]){
+      minx = std::min(minx, float(p.x));
+      maxx = std::max(maxx, float(p.x));
+      minz[view] = std::min(minz[view], float(p.z));
+      maxz[view] = std::max(maxz[view], float(p.z));
+    }
+  }
+
+  // Add some padding
+  for(int view = 0; view < 3; ++view){
+    minz[view] -= 100;
+    maxz[view] += 100;
+  }
+  minx -= 20;
+  maxx += 20;
+
+  // Don't allow the vertex further downstream in z (view 0) than 25% of the
+  // hits.
+  std::vector<float> zs;
+  zs.reserve(pts[0].size());
+  for(const Pt2D& p: pts[0]) zs.push_back(p.z);
+  auto mid = zs.begin() + zs.size()/4;
+  if(mid != zs.end()){
+    std::nth_element(zs.begin(), mid, zs.end());
+    maxz[0] = *mid;
+  }
+
+  std::vector<HeatMap> hms;
+  hms.reserve(3);
+  for(int view = 0; view < 3; ++view){
+    // Approximately cm bins
+    hms.emplace_back(maxz[view]-minz[view], minz[view], maxz[view],
+                     maxx-minx, minx, maxx);
+
+    if(pts[view].empty()) return false;
+
+    std::vector<Line2D> lines;
+    LinesFromPoints(pts[view], lines);
+
+    if(lines.empty()) return false;
+
+    MapFromLines(lines, hms.back());
+  } // end for view
+
+  if(evt_dir){
+    for(int view = 0; view < 3; ++view){
+      view_dirs[view].makeAndRegister<TH2F>("hmap", "", *hms[view].AsTH2());
+    }
+  }
+
+  vtx = FindPeak3D(hms, dirs);
+
+  hms.clear();
+  for(int view = 0; view < 3; ++view){
+    const double x0 = vtx.X();
+    const double z0 = vtx.Dot(dirs[view]);
+
+    std::vector<Line2D> lines;
+    LinesFromPoints(pts[view], lines, z0, x0, 2.5);
+
+    if(lines.empty()) return false; // How does this happen??
+
+    // mm granularity
+    hms.emplace_back(50, z0-2.5, z0+2.5,
+                     50, x0-2.5, x0+2.5);
+
+    MapFromLines(lines, hms.back());
+  }
+
+  vtx = FindPeak3D(hms, dirs);
+
+  if(evt_dir){
+    for(int view = 0; view < 3; ++view){
+      view_dirs[view].makeAndRegister<TH2F>("hmap_zoom", "", *hms[view].AsTH2());
+
+      const double x = vtx.X();
+      const double z = vtx.Dot(dirs[view]);
+      view_dirs[view].makeAndRegister<TGraph>("vtx3d", "", 1, &z, &x);
+    } // end for view
+  } // end if saving plots
+
+  return true;
+}
+
+// ---------------------------------------------------------------------------
+void QuadVtx::produce(art::Event& evt)
+{
+  auto vtxcol = std::make_unique<std::vector<recob::Vertex>>();
+
+  art::Handle<std::vector<recob::Hit>> hits;
+  evt.getByLabel(fHitLabel, hits);
+
+  art::TFileDirectory* evt_dir = fSavePlots ? new art::TFileDirectory(art::ServiceHandle<art::TFileService>()->mkdir(TString::Format("evt%d", evt.event()).Data())) : 0;
+
+  TVector3 vtx;
+  if(FindVtx(*hits, vtx, evt_dir)){
```

Comment:
Two issues:

- At the very least, use a `std::unique_ptr` instead of a bare pointer for exception safety reasons.
- As `evt_dir` is not actually used in this scope, I suggest moving it to `FindVtx` and changing the call to be:

```cpp
if(FindVtx(*hits, vtx, evt.event())) {
  ...
}
```
thus moving the service-handle creation to the `FindVtx` function.

Reply:
Done